In [ ]:
!pip install apyori

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5974 sha256=b96723b4d1f2d9fb24850a19e58df82069c03f4b6dd1ac104c54838fe104b45d
  Stored in directory: /root/.cache/pip/wheels/cb/f6/e1/57973c631d27efd1a2f375bd6a83b2a616c4021f24aab84080
Successfully built apyori


In [ ]:
import numpy as np
import pandas as pd
from apyori import apriori

In [ ]:
store_data = pd.read_csv('https://raw.githubusercontent.com/HuynhHoa99/AprioriAlgorithm/master/Day1.csv', header=None)


In [ ]:
store_data.head()


0      1      2       3     4      5
0  Wine  Chips  Bread  Butter  Milk  Apple
1  Wine    NaN  Bread  Butter  Milk    NaN
2   NaN    NaN  Bread  Butter  Milk    NaN
3   NaN  Chips    NaN     NaN   NaN  Apple
4  Wine  Chips  Bread  Butter  Milk  Apple

In [ ]:
store_data.shape


(22, 6)

In [ ]:
records = []
for i in range(22):
    records.append([str(store_data.values[i, j]) for j in range(6)])

In [ ]:
records


[['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'nan', 'Bread', 'Butter', 'Milk', 'nan'],
 ['nan', 'nan', 'Bread', 'Butter', 'Milk', 'nan'],
 ['nan', 'Chips', 'nan', 'nan', 'nan', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'nan', 'nan', 'Milk', 'nan'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'nan', 'Apple'],
 ['Wine', 'Chips', 'nan', 'nan', 'Milk', 'nan'],
 ['Wine', 'nan', 'Bread', 'nan', 'nan', 'Apple'],
 ['Wine', 'nan', 'Bread', 'Butter', 'Milk', 'nan'],
 ['nan', 'Chips', 'Bread', 'Butter', 'nan', 'Apple'],
 ['Wine', 'nan', 'nan', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'nan'],
 ['Wine', 'nan', 'Bread', 'nan', 'Milk', 'Apple'],
 ['Wine', 'nan', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['nan', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['nan', 'Chips', 'nan', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple

In [ ]:
association_rules = apriori(records, min_support=0.5, min_confidence=0.7, min_lift=1.2, min_length=2)
association_results = list(association_rules)

In [ ]:
print(association_results)


[RelationRecord(items=frozenset({'Bread', 'Butter', 'Milk'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Butter'}), items_add=frozenset({'Bread', 'Milk'}), confidence=0.7333333333333334, lift=1.241025641025641), OrderedStatistic(items_base=frozenset({'Bread', 'Milk'}), items_add=frozenset({'Butter'}), confidence=0.8461538461538461, lift=1.241025641025641)])]


=> Quy luật được phát hiện là những giao dịch mà mua Bread và Milk thì 84.6% cũng mua Butter , và 1 lần Butter được mua không thì sẽ có 1.241 lần Butter sẽ được mua kèm với những người mua bread và milk


Practice with groceries dataset

In [ ]:
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
data = pd.read_excel('http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx')

In [ ]:
data.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [ ]:
data.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [ ]:
data.Country.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [ ]:
data['Description'] = data['Description'].str.strip()

In [ ]:
data.dropna(axis = 0, subset =['InvoiceNo'], inplace = True)
data['InvoiceNo'] = data['InvoiceNo'].astype('str')

In [ ]:
data = data[~data['InvoiceNo'].str.contains('C')]

In [ ]:
# Transactions done in France
basket_France = (data[data['Country'] =="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))
 


In [ ]:
# Transactions done in the United Kingdom
basket_UK = (data[data['Country'] =="United Kingdom"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))


In [ ]:

# Transactions done in Portugal
basket_Por = (data[data['Country'] =="Portugal"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))
 


In [ ]:
basket_Sweden = (data[data['Country'] =="Sweden"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

In [ ]:
def hot_encode(x):
    if(x<= 0):
        return 0
    if(x>= 1):
        return 1

In [ ]:
# Encoding the datasets
basket_encoded = basket_France.applymap(hot_encode)
basket_France = basket_encoded
 
basket_encoded = basket_UK.applymap(hot_encode)
basket_UK = basket_encoded
 
basket_encoded = basket_Por.applymap(hot_encode)
basket_Por = basket_encoded
 
basket_encoded = basket_Sweden.applymap(hot_encode)
basket_Sweden = basket_encoded

In [ ]:
# Building the model
frq_items = apriori(basket_France, min_support = 0.05, use_colnames = True)

France results

In [ ]:
# Collecting the inferred rules in a dataframe
rules = association_rules(frq_items, metric ="lift", min_threshold = 1)
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False])
print(rules.head())

                                           antecedents  ... conviction
44                        (JUMBO BAG WOODLAND ANIMALS)  ...        inf
260  (PLASTERS IN TIN CIRCUS PARADE, RED TOADSTOOL ...  ...        inf
272  (PLASTERS IN TIN WOODLAND ANIMALS, RED TOADSTO...  ...        inf
300  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.897959
301  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.489796

[5 rows x 9 columns]


UK results

In [ ]:
frq_items = apriori(basket_UK, min_support = 0.01, use_colnames = True)
rules = association_rules(frq_items, metric ="lift", min_threshold = 1)
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False])
print(rules.head())

                                       antecedents  ... conviction
117           (BEADED CRYSTAL HEART PINK ON STICK)  ...  39.637371
2020  (SUKI  SHOULDER BAG, JAM MAKING SET PRINTED)  ...  26.096206
2295         (HERB MARKER MINT, HERB MARKER THYME)  ...  21.947227
2300   (HERB MARKER PARSLEY, HERB MARKER ROSEMARY)  ...  20.444951
2301      (HERB MARKER PARSLEY, HERB MARKER THYME)  ...  20.443842

[5 rows x 9 columns]


Portugal results

In [ ]:
frq_items = apriori(basket_Por, min_support = 0.05, use_colnames = True)
rules = association_rules(frq_items, metric ="lift", min_threshold = 1)
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False])
print(rules.head())

                             antecedents  ... conviction
1170    (SET 12 COLOUR PENCILS SPACEBOY)  ...        inf
1171  (SET 12 COLOUR PENCILS DOLLY GIRL)  ...        inf
1172  (SET 12 COLOUR PENCILS DOLLY GIRL)  ...        inf
1173  (SET OF 4 KNICK KNACK TINS LONDON)  ...        inf
1174  (SET 12 COLOUR PENCILS DOLLY GIRL)  ...        inf

[5 rows x 9 columns]


Sweden 

In [ ]:

frq_items = apriori(basket_Sweden, min_support = 0.05, use_colnames = True)
rules = association_rules(frq_items, metric ="lift", min_threshold = 1)
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False])
print(rules.head())

                        antecedents  ... conviction
0     (12 PENCILS SMALL TUBE SKULL)  ...        inf
1     (PACK OF 72 SKULL CAKE CASES)  ...        inf
4    (ASSORTED BOTTLE TOP  MAGNETS)  ...        inf
5           (36 DOILIES DOLLY GIRL)  ...        inf
180  (CHILDRENS CUTLERY DOLLY GIRL)  ...        inf

[5 rows x 9 columns]


Nhận xét kết quả

Eclat 

In [ ]:
!pip install pyECLAT


In [ ]:
transactions = [
    ['beer', 'wine', 'cheese'],
    ['beer', 'potato chips'],
    ['eggs', 'flower', 'butter', 'cheese'],
    ['eggs', 'flower', 'butter', 'beer', 'potato chips'],
    ['wine', 'cheese'],
    ['potato chips'],
    ['eggs', 'flower', 'butter', 'wine', 'cheese'],
    ['eggs', 'flower', 'butter', 'beer', 'potato chips'],
    ['wine', 'beer'],
    ['beer', 'potato chips'],
    ['butter', 'eggs'],
    ['beer', 'potato chips'],
    ['flower', 'eggs'],
    ['beer', 'potato chips'],
    ['eggs', 'flower', 'butter', 'wine', 'cheese'],
    ['beer', 'wine', 'potato chips', 'cheese'],
    ['wine', 'cheese'],
    ['beer', 'potato chips'],
    ['wine', 'cheese'],
    ['beer', 'potato chips']
]

In [ ]:
transactions

[['beer', 'wine', 'cheese'],
 ['beer', 'potato chips'],
 ['eggs', 'flower', 'butter', 'cheese'],
 ['eggs', 'flower', 'butter', 'beer', 'potato chips'],
 ['wine', 'cheese'],
 ['potato chips'],
 ['eggs', 'flower', 'butter', 'wine', 'cheese'],
 ['eggs', 'flower', 'butter', 'beer', 'potato chips'],
 ['wine', 'beer'],
 ['beer', 'potato chips'],
 ['butter', 'eggs'],
 ['beer', 'potato chips'],
 ['flower', 'eggs'],
 ['beer', 'potato chips'],
 ['eggs', 'flower', 'butter', 'wine', 'cheese'],
 ['beer', 'wine', 'potato chips', 'cheese'],
 ['wine', 'cheese'],
 ['beer', 'potato chips'],
 ['wine', 'cheese'],
 ['beer', 'potato chips']]

In [ ]:
import pandas as pd
data = pd.DataFrame(transactions)
data

0             1             2       3             4
0           beer          wine        cheese    None          None
1           beer  potato chips          None    None          None
2           eggs        flower        butter  cheese          None
3           eggs        flower        butter    beer  potato chips
4           wine        cheese          None    None          None
5   potato chips          None          None    None          None
6           eggs        flower        butter    wine        cheese
7           eggs        flower        butter    beer  potato chips
8           wine          beer          None    None          None
9           beer  potato chips          None    None          None
10        butter          eggs          None    None          None
11          beer  potato chips          None    None          None
12        flower          eggs          None    None          None
13          beer  potato chips          None    None          None
14          eggs        flower        butter    wine        cheese
15          beer          wine  potato chips  cheese          None
16          wine        cheese          None    None          None
17          beer  potato chips          None    None          None
18          wine        cheese          None    None          None
19          beer  potato chips          None    None          None

In [ ]:
min_n_products = 2
min_support = 7/len(transactions)
max_length = max([len(x) for x in transactions])


In [ ]:
from pyECLAT import ECLAT


In [ ]:
my_eclat = ECLAT(data=data, verbose=True)


100%|██████████| 8/8 [00:00<00:00, 1290.21it/s]


In [ ]:
rule_indices, rule_supports = my_eclat.fit(min_support=min_support,
                                           min_combination=min_n_products,
                                           max_combination=max_length)

Combination 2 by 2


10it [00:00, 179.29it/s]


Combination 3 by 3


10it [00:00, 178.52it/s]


Combination 4 by 4


5it [00:00, 172.97it/s]


Combination 5 by 5


1it [00:00, 97.05it/s]


In [ ]:
print(rule_supports)


{'potato chips & beer': 0.45, 'wine & cheese': 0.35}
